File: code/ii-service/src/main/java/com/owlunit/service/cinema/MovieService.java
package com.owlunit.service.cinema;

import com.owlunit.model.cinema.Keyword;
import com.owlunit.model.cinema.Movie;
import com.owlunit.model.cinema.Person;
import com.owlunit.model.cinema.Role;
import com.owlunit.service.exception.NotFoundException;

import java.util.Map;

/**
 * @author Ilya Pimenov
 * @author Anton Chebotaev
 *         Owls Proprietary
 */
public interface MovieService {

    Movie createMovie(Movie movie);

    Movie loadByName(String name, Long year) throws NotFoundException;

    Movie loadByExternalId(String service, String externalId) throws NotFoundException;

    Map<Movie, Double> getMostLike(Movie movie) throws NotFoundException;

    Movie setDescription(Movie movie, String description) throws NotFoundException;

    Movie setTranslateName(Movie movie, String translateName) throws NotFoundException;

    Movie setExternalId(Movie movie, String service, String externalId) throws NotFoundException;

    Movie setAkaName(Movie movie, String akaName) throws NotFoundException;

    Movie addPerson(Movie movie, Person person, Role role) throws NotFoundException;

    Movie addKeyword(Movie movie, Keyword keyword) throws NotFoundException;

    Boolean hasKeyword(Movie movie, Keyword keyword) throws NotFoundException;

    Movie addTagline(Movie movie, String tagline) throws NotFoundException;

}

File: code/ii-service/src/main/scala/com/owlunit/service/cinema/MovieService.scala
package com.owlunit.service.cinema

import com.owlunit.core.ii.IiDao
import impl.MovieServiceImpl

/**
 * @author Anton Chebotaev
 *         Owls Proprietary
 */

class MovieIi(override val id: Long,
            val name: String,
            val year: Int,
            val tags: Set[KeywordIi] = Set.empty,
            val persons: Map[Role.Value, Set[PersonIi]] = Map.empty) extends CinemaIi(id, name) {

  def copy(id: Long) = new MovieIi(
    id,
    this.name,
    this.year,
    this.tags,
    this.persons
  )

  override def toString = "MovieIi(%d, %s/%d)" format (id, name, year)

}

object Role extends Enumeration {
  type Role = Value
  val Actor    = Value("Actor")
  val Director = Value("Director")
  val Producer = Value("Producer")
}

trait MovieService {

  def createMovie(sample: MovieIi): MovieIi
  def createMovie(name: String, year: Int): MovieIi
  def loadMovie(id: Long): Option[MovieIi]
  def loadMovie(name: String, year: Int): Option[MovieIi]
  def searchMovie(query: String): Seq[MovieIi]

  def addKeyword(movie: MovieIi, keyword: KeywordIi, startFrequency: Int): MovieIi
  def addPerson(movie: MovieIi, person: PersonIi, newRole: Role.Value): MovieIi

}